cmd_AlterLandscape.lua

```diff
@@ -733,3 +733,233 @@ end
 
 
 
+
+function HandleGenerationShapeCommand(a_Split, a_Player)
+	-- //generate <block> <formula>
+	
+	local IsHollow     = false
+	local UseRawCoords = false
+	local Offset       = false
+	local OffsetCenter = false
+	
+	local NumFlags = 0
+	for Idx, Value in ipairs(a_Split) do
+		NumFlags = NumFlags + 1
+		if (Value == "-h") then
+			IsHollow = true
+		elseif (Value == "-r") then
+			UseRawCoords = true
+		elseif (Value == "-o") then
+			Offset = true
+		elseif (Value == "-c") then
+			OffsetCenter = true
+		else
+			NumFlags = NumFlags - 1
+		end
+	end
+	
+	if ((a_Split[2 + NumFlags] == nil) or (a_Split[3 + NumFlags] == nil)) then
+		a_Player:SendMessage(cChatColor.Rose .. "Too few arguments.")
+		a_Player:SendMessage(cChatColor.Rose .. "//generate [Flags] <block> <formula>")
+		return true
+	end
+	
+	-- Check the selection:
+	local State = GetPlayerState(a_Player)
+	if not(State.Selection:IsValid()) then
+		a_Player:SendMessage(cChatColor.Rose .. "No region set")
+		return true
+	end
+	
+	local BlockTable = RetrieveBlockTypes(a_Split[2 + NumFlags])
+	if not(BlockTable) then
+		a_Player:SendMessage(cChatColor.LightPurple .. "Unknown block type: '" .. a_Split[2 + NumFlags] .. "'.")
+		return true
+	end
+	
+	-- Check with other plugins if the operation is okay:
+	local SrcCuboid = State.Selection:GetSortedCuboid()
+	local World = a_Player:GetWorld()
+	if not(CheckAreaCallbacks(SrcCuboid, a_Player, World, "generation")) then
+		return
+	end
+	
+	-- Push an undo snapshot:
+	State.UndoStack:PushUndoFromCuboid(World, SrcCuboid, "generation")
+	
+	local FormulaString = table.concat(a_Split, " ", 3 + NumFlags)
+	:gsub(";", ",") -- Replace ";" to ",". This makes it possible to set block types/metas in the formula.
+	:gsub("!=", "~=") -- Lua operator for not equal is ~=
+	
+	-- Create a loader for the formula checker. (UNSAFE, FIX USING setfenv)
+	local FormulaLoader = loadstring([[
+	local abs, acos, asin, atan, atan2,
+	ceil, cos, cosh, exp, floor, ln, 
+	log, log10, max, min, round, sin,
+	sinh, sqrt, tan, tanh, random, pi, e
+	=
+	math.abs, math.acos, math.asin, math.atan, math.atan2,
+	math.ceil, math.cos, math.cosh, math.exp, math.floor, math.log,
+	math.log, math.log10, math.max, math.min, Round, math.sin,
+	math.sinh, math.sqrt, math.tan, math.tanh, math.random, math.pi, math.exp(1)
+	
+	-- These functions have no build in function:
+	local cbrt = function(x) return sqrt(x^(1/3)) end
+	local randint = function(max) return random(0, max) end
+	-- TODO: rint function
+
+	return function(x, y, z, type, data)
+		local res = {]] .. FormulaString .. [[}
+		return res[1], res.type or type, res.data or data
+	end]])
+	
+	-- Try to get the formula checker
+	local Succes, Formula = pcall(FormulaLoader)
```

Comment:
This might be quite unsafe. Consider:
`//g block 1,cFile:DeleteFile("settings.ini")`
You should probably use a Lua sandbox instead: http://lua-users.org/wiki/SandBoxes